_posts/2015-11-03-what-you-can-learn-when-you-pair-and-collaborate.md

```diff
@@ -0,0 +1,198 @@
+
+I did the kata under these environmental settings:
+- paired or mobbed with my colleagues at work
+- used pomodoros whilst refactoring and writing tests (both working on my own and with a mob of developers)
+- had a lot of discussions and retrospective after trying different approaches
+
+I will be sharing what I learnt, a technique or two that you can go away with and apply in your daily work or teach it to someone. Here are some refactoring tips and hints to share, that I made notes of whilst doing the kata:
+
+- Refactor the deepest branch first
+- Look for quick wins (removal of magic numbers, naming to reveal intent, etc...)
+- Reorder / reorganise if and else blocks (manually)
+- Reverse if and else blocks without loosing its meaning (automatic IDE refactoring)
+- Combine if blocks with similar conditions (manually)
+- Decompose if-else blocks into simpler if blocks without loosing its meaning
+
+Let's say we apply one or more of the above techniques to the below block of code:
+
+```java
+public void updateQuality(Item[] items) {
+    for (Item item : items) {
+    	// O
+        if (item.name.equals(AGED_BRIE)) {
+            item.increaseQuality();
+        } else if (item.name.equals(BACKSTAGE_PASSES)) {
+            item.increaseQuality();
+
+            if (item.name.equals(BACKSTAGE_PASSES)) {
+                if (item.expiresIn(ELEVEN_DAYS)) {
+                    item.increaseQuality();
+                }
+
+                if (item.expiresIn(SIX_DAYS)) {
+                    item.increaseQuality();
+                }
+            }
+        } else {
+            if (item.name.equals(SULFURAS)) {
+
+            } else {
+                item.decreaseQuality();
+            }
+        }
+
+        // P
+        if (item.name.equals(SULFURAS)) {
+
+        } else {
+            item.decreaseSellIn();
+        }
+
+		// A
+        if (item.name.equals(AGED_BRIE)) {
+            if (item.isExpired()) {
+                item.increaseQuality();
+            }
+        } else {
+        	// B
+            if (item.name.equals(BACKSTAGE_PASSES)) {
+                if (item.isExpired()) {
+                    item.setQualityToMinimum();
+                }
+            } else {
+            	// C
+                if (item.name.equals(SULFURAS)) {
+                    if (item.isExpired()) {
+                        continue;
+                    }
+                } else { // D
+                    if (item.isExpired()) {
+                        item.decreaseQuality();
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+We can see that the code blocks are not all at the same level of abstraction but also there is duplication across them all. All our tests are green so we can move blocks around without fear as we will be alerted if we break existing functionality. To me the above blocks (O, P, ABCD) appear like the below mathematical expression, and can be merged like this expression:
+
```

Comment:
This section will be replaced with a better illustration i.e. picture or a flow diagram.